dependencies.js

```diff
@@ -0,0 +1,412 @@
+ *
+ * Aliases which are not in this map will the get title of the component.
+ * @property {string | string[]} [require]
+ * @property {string | string[]} [modify]
+ * @property {string | string[]} [after]
+ */
+
+var getLoad = (function () {
+
+	/**
+	 * A function which does absolutely nothing.
+	 *
+	 * @type {any}
+	 */
+	var noop = function () { };
+
+	/**
+	 * Converts the given value to an array.
+	 *
+	 * If the given value is already an array, the value itself will be returned.
+	 * `null` and `undefined` will return an empty array.
+	 * For every other value a new array with the given value as its only element will be created.
+	 *
+	 * @param {null | undefined | T | T[]} value
+	 * @returns {T[]}
+	 * @template T
+	 */
+	function toArray(value) {
+		if (Array.isArray(value)) {
+			return value;
+		} else {
+			return value == null ? [] : [value];
+		}
+	}
+
+	/**
+	 * Returns a new set for the given string array.
+	 *
+	 * @param {string[]} array
+	 * @returns {StringSet}
+	 *
+	 * @typedef {Object<string, true>} StringSet
+	 */
+	function toSet(array) {
+		/** @type {StringSet} */
+		var set = {};
+		for (var i = 0, l = array.length; i < l; i++) {
+			set[array[i]] = true;
+		}
+		return set;
+	}
+
+	/**
+	 * Creates a map of every components id to its entry.
+	 *
+	 * @param {Components} components
+	 * @returns {EntryMap}
+	 *
+	 * @typedef {{ [id: string]: Readonly<ComponentEntry> | undefined }} EntryMap
+	 */
+	function createEntryMap(components) {
+		/** @type {EntryMap} */
+		var map = {};
+
+		for (var categoryName in components) {
+			var category = components[categoryName];
+			for (var id in category) {
+				if (id != 'meta') {
+					/** @type {ComponentEntry | string} */
+					var entry = category[id];
+					map[id] = typeof entry == 'string' ? { title: entry } : entry;
+				}
+			}
+		}
+
+		return map;
+	}
+
+	/**
+	 * Creates a full dependencies map which includes all types of dependencies and their transitive dependencies.
+	 *
+	 * @param {EntryMap} entryMap
+	 * @returns {DependencyMap}
+	 *
+	 * @typedef {Object<string, StringSet>} DependencyMap
+	 */
+	function createDependencyMap(entryMap) {
+		/** @type {DependencyMap} */
+		var map = {};
+
+		/**
+		 * Adds the dependencies of the given component to the dependency map.
+		 *
+		 * @param {string} id
+		 * @param {string[]} stack
+		 */
+		function addToMap(id, stack) {
+			if (id in map) {
+				return;
+			}
+
+			stack.push(id);
+
+			// check for circular dependencies
+			var firstIndex = stack.indexOf(id);
+			if (firstIndex < stack.length - 1) {
+				throw new Error('Circular dependency: ' + stack.slice(firstIndex).join(' -> '));
+			}
+
+			/** @type {StringSet} */
+			var dependencies = {};
+
+			var entry = entryMap[id];
+			if (entry) {
+				/** @type {string[]} */
+				var deps = (/** @type {any[]} */([]).concat(entry.require, entry.modify, entry.after).filter(Boolean));
```

Comment:
Is this weird `@type any[]` thing necessary? Makes the code hard to read (esp w/ the extra parens).

Reply:
I know it sucks but I couldn't find another way to make TypeScript happy.

Let's keep this open here. I'll try to find a better solution in a bit.

Reply:
Correction: Seems like they fixed this. I'll just remove the `/** @type {any[]} */` plus parens.